pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -456,44 +620,66 @@ impl InMemoryLayer {
         })
     }
 
-    // Write path.
+    /// Write path.
+    ///
+    /// Errors are not retryable, the [`InMemoryLayer`] must be discarded, and not be read from.
+    /// The reason why it's not retryable is that the [`EphemeralFile`] writes are not retryable.
+    /// TODO: it can be made retryable if we aborted the process on EphemeralFile write errors.
     pub async fn put_batch(
         &self,
         serialized_batch: SerializedBatch,
         ctx: &RequestContext,
-    ) -> Result<()> {
+    ) -> anyhow::Result<()> {
         let mut inner = self.inner.write().await;
         self.assert_writable();
 
-        let base_off = {
-            inner
-                .file
-                .write_raw(
-                    &serialized_batch.raw,
-                    &RequestContextBuilder::extend(ctx)
-                        .page_content_kind(PageContentKind::InMemoryLayer)
-                        .build(),
-                )
-                .await?
-        };
-
-        for SerializedBatchOffset {
-            key,
-            lsn,
-            offset: relative_off,
-        } in serialized_batch.offsets
-        {
-            let off = base_off + relative_off;
+        let base_offset = inner.file.len();
+
+        // Add the base_offset to the batch's index values which are relative to the batch start.
+        let index_values: Vec<SerializedBatchOffset> = serialized_batch
+            .offsets
+            .into_iter()
+            .map(|SerializedBatchOffset { key, lsn, value }| {
+                let InMemoryLayerIndexValueUnpacked {
+                    will_init,
+                    len,
+                    pos,
+                } = value.unpack();
+                anyhow::Ok(SerializedBatchOffset {
+                    key,
+                    lsn,
+                    value: InMemoryLayerIndexValue::new(InMemoryLayerIndexValueNewArgs {
+                        base_offset,
+                        batch_offset: pos,
+                        len: len.into_usize(),
+                        will_init,
+                    })?,
+                })
+            })
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        // Write the batch to the file
+        inner.file.write_raw(&serialized_batch.raw, ctx).await?;
+        let new_size = inner.file.len();
+        let expected_new_len = base_offset
+            .checked_add(serialized_batch.raw.len().into_u64())
+            // write_raw would error if we were to overflow u64.
+            // also InMemoryLayerIndexValue and higher levels in
+            //the code don't allow the file to grow that large
+            .unwrap();
+        assert_eq!(new_size, expected_new_len);
+
+        // Update the index with the new values
+        for SerializedBatchOffset { key, lsn, value } in index_values {
```

Comment:
nit: `value` here is confusing, it reads like it's a Value rather than the size metadata

Reply:
35b81a0a034bbdd655877fc2ab552534a5494694